src/main/java/org/optaplanner/springboottaskassigning/solver/DefaultSolverManager.java

```diff
@@ -103,17 +125,39 @@ public void addEventListener(Object problemId, SolverEventListener<Solution_> ev
     @Override
     public void shutdown() {
         logger.info("Shutting down {}.", DefaultSolverManager.class.getName());
-        executorService.shutdown();
+        stopSolvers();
+        solverExecutorService.shutdown();
+        eventHandlerExecutorService.shutdown();
         Long awaitingDuration = 1L;
         try {
-            if (!executorService.awaitTermination(awaitingDuration, TimeUnit.SECONDS)) {
+            if (!solverExecutorService.awaitTermination(awaitingDuration, TimeUnit.SECONDS)
+                    || !eventHandlerExecutorService.awaitTermination(awaitingDuration, TimeUnit.SECONDS)) {
                 logger.info("Still waiting shutdown after {} second, calling shutdownNow().", awaitingDuration);
             }
         } catch (InterruptedException e) {
             Thread.currentThread().interrupt();
             throw new IllegalStateException("SolverManager thread interrupted while awaiting termination.", e);
         } finally {
-            executorService.shutdownNow();
+            solverExecutorService.shutdownNow();
+            eventHandlerExecutorService.shutdownNow();
+        }
+    }
+
+    private void stopSolvers() {
+        for (SolverTask solverTask : problemIdToSolverTaskMap.values()) {
+            stopSolverTask(solverTask);
+        }
+    }
+
+    private void stopSolverTask(SolverTask solverTask) {
+        solverTask.stopSolver();
+        try {
+            problemIdToCompletableFutureMap.get(solverTask.getProblemId()).get();
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new RuntimeException("Error in solver (" + solverTask.getProblemId() + ").", e);
+        } catch (ExecutionException e) {
+            throw new RuntimeException("Error in solver (" + solverTask.getProblemId() + ").", e);
```

Comment:
It is a wrapper, but re-throwing it instead of `e.getCause()` gives more info about the thread in which the exception occurred instead of only the stack trace starting from `solverTask.call()` (or whatever method is passed to the Executor).

Actually now that [`solverFuture.handle()`](https://github.com/MusaTalluzi/spring-boot-task-assigning/pull/2/files/e7302ca261a7bebfc8e7a10812c2ae8fff50bdf5#diff-8b7789ac378a5aacdfb02dfc865db450R74) handles exceptions there's no need to catch them here an re-throw them.

Reply:
Le's add some test for exception propagation and if it works like that => +1 for simplifying it here.